ChatRPG/ChatRPG/Pages/Campaign.razor

```diff
@@ -0,0 +1,94 @@
+﻿@page "/Campaign"
+@using ChatRPG.Data
+@using ChatRPG.API
+@inject IOpenAiLlmClient OpenAiLlmClient;
+@inject IJSRuntime JsRuntime
+
+<div class="text-center mx-auto">
+    <h1>Campaign</h1>
+</div>
+
+<div class="conversation mx-auto col-8 mt-4 rounded p-5 text-black">
+    <div class="conversation-text">
+        @foreach (string message in _conversation)
+        {
+            @if (message.StartsWith("Player: "))
+            {
+                <div class="user-message-container">
+                    <p class="user-message">@message</p>
+                </div>
+            }
+            else
+            {
+                <div class="assistant-message-container">
+                    <p class="assistant-message">@message</p>
+                </div>
+            }
+        }
+    </div>
+</div>
+
+<div class="container">
+    <div class="input-group mx-auto col-8 mt-2 mb-5 input-group-border">
+        <input type="text" style="resize:vertical;" class="form-control user-prompt custom-text-field"
+               @bind="_userInput" placeholder="What do you do?" @onkeyup="@EnterKeyHandler"/>
+        <div class="input-group-append">
+            <button class="btn btn-primary ml-2" type="button" @onclick="SendPrompt">
+                Send
+            </button>
+        </div>
+    </div>
+</div>
+
+<span id="bottom-id"></span>
+
+@code {
+    private readonly FileUtility _fileUtil = new FileUtility();
+    readonly List<string> _conversation = new List<string>();
+    private string _userInput = "";
+
+    async Task EnterKeyHandler(KeyboardEventArgs e)
+    {
+        if (e.Code is "Enter" or "NumpadEnter")
+        {
+            await SendPrompt();
+        }
+    }
+
+    async Task SendPrompt()
+    {
+        if (!string.IsNullOrWhiteSpace(_userInput))
+        {
+            // Use the user input to generate a response
+            _conversation.Add($"Player: {_userInput}");
+            string response = await ProcessUserInput(_userInput);
+            _conversation.Add($"Assistant: {response}");
+            await _fileUtil.UpdateSaveFileAsync(new MessagePair(_userInput, response));
+
+            // Clear the user input
+            _userInput = "";
+
+            // Notify the component to re-render because of async methods
+            StateHasChanged();
+
+            // Scroll to bottom of page
+            await ScrollToElement("bottom-id");
+        }
+    }
+
+    async Task<string> ProcessUserInput(string input)
+    {
+        List<OpenAiGptInputMessage> inputMessage = new List<OpenAiGptInputMessage> { new OpenAiGptInputMessage("user", input) };
+
+        ChatCompletionObject response = await OpenAiLlmClient.GetChatCompletion(inputMessage);
+
+        return response.Choices[0].Message.Content;
+    }
+
+    async Task ScrollToElement(string elementId)
+    {
+        IJSObjectReference module = await JsRuntime.InvokeAsync<IJSObjectReference>("import", "./js/scroll.js");
```

Comment:
Er det muligt at importere `scroll.js` én gang, evt. når viewet initialiseres?

Reply:
Jeg brugte alt for længe på det og fik det aldrig til at virke. Den skulle gerne kunne importere alt der ligger i wwwroot/js automatisk så man kan bruge det med JSRuntime, men det ville ikke virke. Så jeg fandt den her workaround og det burde være det eneste sted at det script skal bruges. Hvis du mener om scriptet bare kan loades og gemmes i en private variable, så er svaret ja haha

Reply:
nvm, that's a negative on private variables da måden man henter dem ind med JSRuntime er asyncronous og det kan ikke gøres statisk. Det er dog en lokal fil, så tidsmæssigt er det ret hurtigt. Jeg kan godt bruge forsøge at finde ud af om jeg kan få den til at loade det automatisk, men det virkede ikke før.

Reply:
Ah I see. 
Ja, jeg mener om man kan flytte det her kald ud i en initialiserings-funktion (eller måske er det bare en privat variable ligesom fx. `_conversation`, så godt kender jeg ikke til Razor).
Jeg forestiller mig det er dyrt at invokere JS sådan her, men med den her ændring kan vi undgå at importere scriptet hver gang det skal bruges.

Reply:
Det er ikke dyrt da det er en lokal fil, så hvis ikke jeg finder en løsning, så gør det ikke så meget.

Reply:
@KarmaKamikaze Har du kigget [på det her?](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?view=aspnetcore-7.0#component-initialization-oninitializedasync)

Reply:
> @KarmaKamikaze Har du kigget [på det her?](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?view=aspnetcore-7.0#component-initialization-oninitializedasync)

Yep, all afternoon, buddy. Det virkede ikke selvom det burde.

Reply:
@damniko nvm, det virkede sgu! Jeg ved ikke hvad jeg gjorde forkert før, men jeg sværger jeg havde prøvet det her!

Reply:
Jeg gætter på at det var @Sarmisuper der trollede, men fedt at det virker nu!

Reply:
![Uploading image.png…]()